### PR TITLE
[iOS] send images and videos to Photos with moveToSharedStorage

### PR DESCRIPTION
**Pull request by wreppun**
TL;DR; Needed this for myself, so thought I'd share

## Background
On iOS, when a user downloads an image or video, they generally expect it to show up in their camera roll.  Since this package already exposes a `SharedStorage` enum, including `.images` and `.video`, I implemented this on the iOS side.

## This Breaks Things
This implementation breaks `pathInSharedStorage`, which is designed to take the initial filename, and extrapolate the shared path/filename.  Since iOS creates a new filename for images/videos moved to Photos, this method cannot be implemented AFAIK.

But:
1. I didn't realize this broke `pathInSharedStorage` until after I had implemented this PR
2. I couldn't actually figure out a use case for `pathInSharedStorage` (if the user moves the file from its original location, surely they should be responsible for tracking the new location?)

## Warning
I don't really write swift.  There were a lot of google searches like "swift unwrap optionals" and "swift anonymous function examples".  And I am still generally confused by their API versioning.  So this code could be terrible?

### CODE REVIEW

**Comment by 781flyingdutchman** [comment]
Hi, thanks so much for creating this PR!  Not being an iPhone user myself I've always wondered if something like this was possible, but never found the time to investigate. Thank you!  A few thoughts on this:
* I think the fact that this breaks `pathInSharedStorage` is not a problem, as this is a special case where the OS determines the path, and - if I understand correctly - you do return the new path from `moveToSharedStorage` for videos and photos
* The bigger breaking issue is that this changes the functionality of `moveToSharedStorage` (in a good way, but unexpected if you upgrade the package and didn't intend this to happen). I think we can fix that by adding a flag `useiOSPhotosLibrary` to the `moveToSharedStorage` call that uses your behavior when set to true, but the old behavior when set to false (default)
* I don't see any code about getting permissions, but when I (now) read the docs for the photos app I see references to [requesting authorization](https://developer.apple.com/documentation/photokit/phphotolibrary/3616052-authorizationstatus) (also for iOS versions lower than 14, which we need as we support 13, [here](https://developer.apple.com/documentation/photokit/phphotolibrary/1620745-authorizationstatus). This would have to be added in the `moveToSharedStorage` method (i.e. check for auth, request if not yet given, only then execute the move)
* I'll add a few comments/suggestions to your PR commit - let me know if you have the time/appetite to address those (which I would be grateful for)
* I will fork your repo/branch and may suggest a few changes to the Swift code

Again, thanks much for your help!

File: ios/Classes/Downloader.swift [hunk 1]
```diff
@@ -384,7 +384,12 @@ public class Downloader: NSObject, FlutterPlugin, URLSessionDelegate, URLSession
             result(nil)
             return
         }
-        result(moveToSharedStorage(filePath: filePath, destination: destination, directory: directory))
+        moveToSharedStorage(
```
Comment (781flyingdutchman) [inline ios/Classes/Downloader.swift:387]: To make this simpler and more consistent with how we've handled the result, I suggested adding a parameter `result: @escaping FlutterResult` to the new `moveToSharedStorage` function. From there you can then simply call `result(yourPath)` or `result(nil)` to pass the result back to Flutter.  You do then need to follow that statement with `return` to exit the method.
What you have done here is adding another `completer` layer to the process, which makes the code a bit hard to read/follow

File: ios/Classes/SharedStorage.swift [hunk 2]
```diff
@@ -18,38 +19,84 @@ public enum SharedStorage: Int {
 }
 
 /// Move file at [filePath] to shared storage [destination] with optional [directory]
-public func moveToSharedStorage(filePath: String, destination: SharedStorage, directory: String) -> String? {
```
Comment (781flyingdutchman) [inline ios/Classes/SharedStorage.swift:21]: Shouldn't you check for the existence of the file in add circumstances, i.e. whether you use your images/photos approach or the old 'fake' one? I think it's best to leave that check here, and return if the file doesn't exist

File: ios/Classes/SharedStorage.swift [hunk 3]
```diff
@@ -18,38 +19,84 @@ public enum SharedStorage: Int {
...
-        return nil
-    }
-    let fileUrl = NSURL(fileURLWithPath: filePath)
-    guard let directory = try? directoryForSharedStorage(destination: destination, directory: directory) else {
-        os_log("Cannot move to shared storage: no permission for directory %@", log: log, type: .error, directory)
-        return nil
+public func moveToSharedStorage(filePath: String, destination: SharedStorage, directory: String, complete: @escaping (String?) -> Void) {
+    switch destination {
+    case .images:
+        saveImageToPhotos(filePath: filePath, complete: complete)
```
Comment (781flyingdutchman) [inline ios/Classes/SharedStorage.swift:25]: There is a lot of code duplication between `saveVideoToPhotos` and `saveImageToPhotos`. I would suggest creating one method `saveToPhotos(filePath: filePath, destination: SharedStorage, complete: complete)` and use the `destination` parameter to reflect the subtle changes between videos and photos, eg `PHAssetChangeRequest.creationRequestForAssetFromVideo` vs `PHAssetChangeRequest.creationRequestForAssetFromImage`

File: ios/Classes/SharedStorage.swift [hunk 4]
```diff
@@ -18,38 +19,84 @@ public enum SharedStorage: Int {
...
-            try FileManager.default.createDirectory(at: directory, withIntermediateDirectories:  true)
-        } catch {
-            os_log("Cannot move to shared storage: failed to create directory %@: %@", log: log, type: .error, directory.path, error.localizedDescription)
-            return nil
+}
+
+private func saveImageToPhotos(filePath: String, complete: @escaping (String?) -> Void) {
+    var localId: String?
+
+    PHPhotoLibrary.shared().performChanges({
```
Comment (781flyingdutchman) [inline ios/Classes/SharedStorage.swift:36]: I believe that before you `performChanges` you need to check for authorization to do so, using:
* [autohorizationStatus](https://developer.apple.com/documentation/photokit/phphotolibrary/3616052-authorizationstatus) and [requestAuthorization](https://developer.apple.com/documentation/photokit/phphotolibrary/3616053-requestauthorization) for iOS 14+
* [autohorizationStatus](https://developer.apple.com/documentation/photokit/phphotolibrary/1620745-authorizationstatus) and [requestAuthorization](https://developer.apple.com/documentation/photokit/phphotolibrary/1620736-requestauthorization) for <iOS14

The `requestAuthorization` calls require a handler that is called once the authorization has been obtained (or not), so you should move the `performChanges` block to execute in that handler, after confirming you have the appropriate `PHAuthorizationStatus`

File: ios/Classes/Downloader.swift
Comment (781flyingdutchman) [inline ios/Classes/Downloader.swift:1]: To make this change backward compatible, we need to add a bool `useIOSPhotoLibrary` to the `moveToSharedStorage` call in Flutter (with `false` as the default), and pass that as the 4th argument to the native code in the `NativeDownbloader` code. In the function `methodMoveToSharedStorage` you can extract the boolean as `let useIOSPhotoLibrary = args[3] as? Bool`, pass it to the `moveToSharedStorage` function and use your new code only if that value is true, otherwise use the `fake` version.  In a future major version release we can change that, but for now we need to maintain backward compatibility

Comment (781flyingdutchman) [reply ios/Classes/Downloader.swift:1]: Actually, I'm wondering if this isn't a good opportunity to make a major version change (so we don't need to be backward compatible). Let's hold off on adding this `useIOSPhotoLibrary`.

File: ios/Classes/Downloader.swift (hunk 1 above)
Comment (wreppun) [reply ios/Classes/Downloader.swift:387]: Yeah.  Several methods need to use callbacks here.

I was trying to prevent the FlutterResult from leaking to peripheral modules, but that's easy to change.

File: ios/Classes/SharedStorage.swift (hunk 4 above)
Comment (wreppun) [reply ios/Classes/SharedStorage.swift:36]: This is something I wasn't super sure on.  Do we want to leave this responsibility outside of this library (e.g. ask the client to use the `permission_handler` library before calling `moveToSharedStorage`)?

If you want to request authorization from _within_ this library, I think the proper move would be to refactor and handle any authorization rejections that come up, and pass those along to the calling function (on the flutter side).  Otherwise, there's no way for the calling library to know when to retry/abort/show-a-notification-about-changing-permissions-via-settings.

File: ios/Classes/SharedStorage.swift (hunk 3 above)
Comment (wreppun) [reply ios/Classes/SharedStorage.swift:25]: Yeah, would you support narrowing the enum here, to SupportedDestinations?  I just didn't want to either:
- pass through an enum where 4/6 cases are not used
- use composition to handle the different retrieval of the change request and the retrieval of the final url

Comment (781flyingdutchman) [reply ios/Classes/SharedStorage.swift:25]: I think it can be done with the existing enum `SharedStorage` and just using an if-statement where required, e.g. 
```
let assetChangeRequest = destination == SharedStorage.video ? 
    PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: fileURL) :
    PHAssetChangeRequest.creationRequestForAssetFromImage(atFileURL: fileURL)
// continue as you have with assetChangeRequest
```

File: ios/Classes/SharedStorage.swift (hunk 4 above)
Comment (781flyingdutchman) [reply ios/Classes/SharedStorage.swift:36]: Yes, that is an issue. We haev teh same issue with notification permissions, and the choice I've made there is to make the authorization request the first time it is required. You are correct that if the user declines the permission, there is no programmatic way to ask for it again (the plugin will assume that decision is permanent, and not show notifications unless the user goes to settings and manually grants the permission).

While not perfect, this is easiest and reasonable, IMHO.  If the developer wants the option to ask again, then they should use a permission management package (several of which exist), and the result of those permissions do apply to the background_downloader, so there is an option for developers that need it.

For the photos library I think we should do the same: ask for permission if needed (i.e. if not already granted or denied) and return `null` from `moveToSharedStorage` if the permission is denied (and log something to that effect).  Importantly, this is also consistent with how it works on Android.

**Comment by 781flyingdutchman** [comment]
Since you appear to be familiar with iOS, I have a related question. Currently, the `.downloads` destination is also 'faked' (just like the .images and .video ones) and I have not been able to figure out how to get permission to add to the iPhone's Downloads folder such that the file shows up in the Files app.  I understand some apps, like Safari but also Chrome, are able to do this, but I have not figured out how to do this in this package.  It would be great if we could fix that issue here too, at the same time.  Do you have any ideas?

File: ios/Classes/SharedStorage.swift (hunk 4 above)
Comment (wreppun) [reply ios/Classes/SharedStorage.swift:36]: First: sure, this is your library, and it's up to you.

But: As soon as a consumer of this library has the intention of putting this into production, then the decline flow _needs_ to be handled.  The user experience is brutal otherwise.  And this just returns `null`.  No message, and nothing actionable.

I know, because I've gone through the swift code, that this is the case.  But this is opaque to the library consumer.

IMO better to have zero permissions handling than half-baked permissions handling.

Comment (wreppun) [reply ios/Classes/SharedStorage.swift:36]: > IMO better to have zero permissions handling than half-baked permissions handling.

Because it makes it clear upfront to the consumer of the library what is needed.

**Comment by wreppun** [comment]
> Since you appear to be familiar with iOS, I have a related question. Currently, the .downloads destination is also 'faked' (just like the .images and .video ones) and I have not been able to figure out how to get permission to add to the iPhone's Downloads folder such that the file shows up in the Files app.

I'm pretty sure I started on this last week by searching "swift move image to camera roll" :)

So I don't have any native knowledge of iOS (or even swift syntax!).  We just needed this for work.

Just poking around on the simulator (I'm on Android myself), I do see the Files app, but there isn't a Downloads folder in there (but maybe because nothing has been downloaded?).

File: ios/Classes/SharedStorage.swift (hunk 4 above)
Comment (wreppun) [reply ios/Classes/SharedStorage.swift:36]: Re-reading this, it came off a lot firmer than I intended.  I do understand that you have already established a pattern of how `background_downloader` handles permissions, and I'm happy to follow it.  I just wanted to share some of my thoughts about potential pitfalls.

**Comment by 781flyingdutchman** [comment]
Yes, what does work is making files in the app's Documents folder visible in the Files app. That takes some permissions and is explained in the README. What I can't get to work is for something to show up in the Files app under Downloads, which is where downloads done by Chrome and Safari end up - so I was just wondering if you are aware of how that can be done.

File: ios/Classes/SharedStorage.swift (hunk 4 above)
Comment (781flyingdutchman) [reply ios/Classes/SharedStorage.swift:36]: I understand what you're saying. One option is to add methods to the FileDownloader to manage permissions directly. For example, you could use one method to get the status of a certain permission, and another to ask for permission.  It gets a bit tricky in that it needs to work for both iOS and Android, and of course the permissions aren't exactly the same. 

I'm hesitant because I worry I'll end up re-implementing what permission plugins already do well today. My question to you therefore is: if you need to manage a user declining certain permissions, then why don't you use one of those permission plugins? And if you don't need to manage the decline (e.g. I don't think you need to manage a user declining your download notifications) then the background_downloader plugin does everything you need.  Isn't that a reasonable approach? I am genuinely trying to find the right approach here, so want to understand your perspective better.
